base/commands/serializer/schema.go

```diff
@@ -0,0 +1,188 @@
+package serializer
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type Schema struct {
+	Imports    []string
+	Namespace  string
+	Classes    []Class
+	ClassNames map[string]Class
+}
+
+type Class struct {
+	Name      string
+	Fields    []Field
+	Namespace string
+}
+
+type Field struct {
+	Name     string
+	Type     string
+	External bool
+}
+
+type ClassInfo struct {
+	FileName  string
+	ClassName string
+	Namespace string
+}
+
+func readSchema(path string) ([]byte, error) {
+	s, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("can not read schema %s", err.Error())
+	}
+	return s, nil
+}
+
+func parseSchema(schema []byte) (Schema, error) {
+	schemaMap, err := YAMLToMap(schema)
+	if err != nil {
+		return Schema{}, err
+	}
+	isValid, schemaErrors, err := validateWithJSONSchema(schemaMap)
+	if err != nil {
+		return Schema{}, err
+	}
+	if !isValid {
+		return Schema{}, fmt.Errorf("schema is not valid, validation errors:\n%s", strings.Join(schemaErrors, "\n"))
+	}
+	sch, err := convertMapToSchema(schemaMap)
+	if err != nil {
+		return Schema{}, err
+	}
+	return sch, nil
+}
+
+func convertMapToSchema(schemaMap map[string]interface{}) (Schema, error) {
+	var schema Schema
+	if err := transcode(schemaMap, &schema); err != nil {
+		return Schema{}, err
+	}
+	return schema, nil
+}
+
+func transcode(in, out interface{}) error {
+	buf := new(bytes.Buffer)
+	if err := json.NewEncoder(buf).Encode(in); err != nil {
+		return err
+	}
+	return json.NewDecoder(buf).Decode(out)
+}
+
+func processSchema(schemaPath string, schema *Schema) error {
+	schema.ClassNames = make(map[string]Class)
+	err := registerClasses(*schema, schema.ClassNames)
+	if err != nil {
+		return err
+	}
+	err = processImports(*schema, schemaPath, schema.Imports, map[string]struct{}{schemaPath: {}})
+	if err != nil {
+		return err
+	}
+	err = checkFieldTypes(*schema, schema.ClassNames)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func processImports(mainSchema Schema, schemaDir string, importPaths []string, importedPaths map[string]struct{}) error {
```

Comment:
I will send a separate PR when I have the time to refactor the rest of the code in this file.